clients/cobol-lsp-vscode-extension/src/services/ControlFlowService.ts

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2025 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+import * as vscode from "vscode";
+import { Program } from "@code4z/analysis/lib/model/cfast";
+import { ControlFlowGraphBuilder } from "@code4z/analysis/lib/graphbuilder";
+import { Graph } from "@code4z/analysis/lib/model/Graph";
+import { SettingsService } from "./Settings";
+
+/**
+ * Control Flow Analysis callback
+ */
+export interface ControlFlowAnalysisCallback {
+  (graphs: Graph[]): void;
+}
+
+/**
+ * Adds a Control Flow Analysis callback function, this callback will be fired on analysis finished event
+ * @param documentUri the document uri
+ * @param callback a callback, that will be fired on finish analysis
+ */
+export function addControlFlowAnalysisCallback(
+  documentUri: string,
+  callback: ControlFlowAnalysisCallback,
+): void {
+  ControlFlowAnalysisService.instance().addCallback(documentUri, callback);
+}
+
+/**
+ * Handle AST and starts analysis it
+ */
+export function controlFlowAstHandler(result: ApiResult): void {
+  if (result.documentUri) {
+    ControlFlowAnalysisService.instance().cancelAnalysis(result.documentUri);
+    if (result.controlFlowAST.length > 0) {
+      ControlFlowAnalysisService.instance().queueAnalysis(
+        result.controlFlowAST,
+        result.documentUri,
+      );
+    }
+  }
+}
+
+export class ApiResult {
+  public controlFlowAST: Program[] = [];
+  public documentUri: string | undefined;
+}
+
+interface AnalysisServiceDelegate {
+  finishTask(
+    documentUri: string,
+    graphs: Graph[],
+    diagnostics: Map<string, vscode.Diagnostic[]>,
+  ): void;
+}
+
+class AnalysisTask {
+  private controller: AbortController;
+  private promise: Promise<Graph[]>;
+
+  constructor(
+    private documentUri: string,
+    public programs: Program[],
+    private delegate: AnalysisServiceDelegate,
+  ) {
+    this.controller = new AbortController();
+    const cfgBuilder = new ControlFlowGraphBuilder(
+      SettingsService.getMaxVMCount(),
+      SettingsService.getUnreachableCodeSeverity(),
+    );
+
+    this.promise = new Promise<Graph[]>((_resolve, reject) => {
+      setImmediate(() => {
+        const abortHandler = () => {
+          reject(new Error("Aborted"));
+        };
+        this.controller.signal.addEventListener("abort", abortHandler);
+        const result = cfgBuilder.build(programs);
```

Comment:
If any of this code throws, the promise will be left unresolved as well.

Also the `build` runs synchronously, so it does not particularly help with the responsiveness problem.